drivers/include/periph/i2c.h

```diff
@@ -120,167 +143,126 @@ typedef enum {
  * @return                  -1 on undefined device given
- * @return                  -2 on invalid address
  */
-int i2c_read_byte(i2c_t dev, uint8_t address, char *data);
+int i2c_init(i2c_t dev);
 
 /**
- * @brief   Read multiple bytes from an I2C device with the given address
+ * @brief   Get mutually exclusive access to the given I2C bus
  *
- * @param[in]  dev          I2C peripheral device
- * @param[in]  address      bus address of the target device
- * @param[out] data         array holding the received bytes
- * @param[in]  length       the number of bytes to read into `data`
+ * In case the I2C device is busy, this function will block until the bus is
+ * free again.
  *
- * @return                  the number of bytes that were read
- * @return                  -1 on undefined device given
+ * @param[in] dev           I2C device to access
+ *
+ * @return                  0 on success
  */
-int i2c_read_bytes(i2c_t dev, uint8_t address, char *data, int length);
+int i2c_acquire(i2c_t dev);
 
 /**
- * @brief   Read one byte from a register at the I2C slave with the given
- *          address
- *
- * @param[in]  dev          I2C peripheral device
- * @param[in]  address      bus address of the target device
- * @param[in]  reg          the register address on the targeted I2C device
- * @param[out] data         the result that was read
+ * @brief   Release the given I2C device to be used by others
  *
- * @return                  the number of bytes that were read
- * @return                  -1 on undefined device given
+ * @param[in] dev           I2C device to release
  */
-int i2c_read_reg(i2c_t dev, uint8_t address, uint8_t reg, char *data);
+void i2c_release(i2c_t dev);
 
 /**
- * @brief   Read multiple bytes from a register at the I2C slave with the given
- *          address
+ * @brief   Read or write data from/to the given I2C device
  *
  * @param[in]  dev          I2C peripheral device
- * @param[in]  address      bus address of the target device
- * @param[in]  reg          the register address on the targeted I2C device
+ * @param[in]  addr         7-bit or 10-bit device address (right-aligned)
  * @param[out] data         array holding the received bytes
- * @param[in]  length       the number of bytes to read into `data`
+ * @param[in]  len          the number of bytes to read into @p data
+ * @param[in]  flags        flags to specify read/write, no stop or addr width
  *
- * @return                  the number of bytes that were read
- * @return                  -1 on undefined device given
+ * @return                  I2C_ACK on successful transfer of @p len byte
+ * @return                  I2C_ADDR_NACK if response to address byte was NACK
+ * @return                  I2C_ERR for any other error
  */
-int i2c_read_regs(i2c_t dev, uint8_t address, uint8_t reg,
-                  char *data, int length);
+int i2c_transfer(i2c_t dev, uint16_t addr,
+                 uint8_t *data, size_t len, uint8_t flags);
```

Comment:
`data` should be `void *`
I think it would make sense to not combine read and write in the same API call. Implementation wise there will always be a difference between the read and the write calls which will end up in this function simply containing two big branches, one for read and one for write.
Also, with a separate write function it is possible to let the buffer be `const` and keep `*data` in program flash without copying to a RAM buffer first.